### DISCUSSION TITLE
Variables unique indices ordering

### DISCUSSION THREAD

**Discussion by th-yoon** [ds4dm/ecole#237] [Questions] [answered]
Hi, I have a question about variables unique indices ordering in the original problem space and in the transformed problem space.
Is the ordering of variables unique indices preserved in the transformed problem space as well?
Thanks.

**Comment by gasse** [comment, marked as answer]
Hi @th-yoon ,

SCIP keeps track of two problems internally:
- the original problem ([`scip->origprob`](https://www.scipopt.org/doc/html/struct__scip_8h_source.php#l00071))
- the transformed problem ([`scip->transprob`](https://www.scipopt.org/doc/html/struct__scip_8h_source.php#l00089))

Each problem then contains, among other things, a list of variables ([`prob->vars`](https://www.scipopt.org/doc/html/struct__prob_8h_source.php#l00055) array). The index of a variable then directly indicates the position of that variable in the array where it is stored, and variable are problem-specific (a variable belongs to one problem only). Hence, variable indices are unique only within a specific problem.

In general, I think most variables from the original problem (say, `x25`) will have an equivalent transformed variable in the transformed problem (`t_x25`), and probably the indices of the variables will be correlated. But in general you can not assume their indices will match. Also, some original variables might get fixed or aggregated in the transformed problem, and thus will appear in [`prob->fixedvars`](https://www.scipopt.org/doc/html/struct__prob_8h_source.php#l00056).

I believe SCIP has a way to map variables to/from the original and transformed spaces (see for example [`SCIPvarGetOrigvarSum()`](https://www.scipopt.org/doc/html/group__PublicVariableMethods.php#ga43bc31c19096e1f77959dab2b594fec1)), although I am not very familiar with that part of the API. In general, I would advise to stick to the transformed problem space to avoid complications.

I hope that helps,

Best,
Maxime

**Reply by th-yoon** [reply]
Thanks for your answer @gasse . One follow-up question is: If I save the transformed problem `A.lp` from the original problem `A.cip`, then would it return the transformed variables of `A.cip` from calling `getVars(transformed=False)` on `A.lp`?

**Reply by gasse** [reply]
I think so yes. See also this piece of code which does essentially that, but directly with SCIP copies without the writing to files part:
https://github.com/ds4dm/ml4co-competition/discussions/31#discussioncomment-1043818

**Reply by th-yoon** [reply]
Thanks!
